packages/core/src/mesh/PrimitiveMesh.ts

```diff
@@ -670,6 +670,180 @@ export class PrimitiveMesh {
+    const indices = PrimitiveMesh._generateIndices(engine, totalVertexCount, totalRectangleCount * 6);
+
+    const positions: Vector3[] = new Array(totalVertexCount);
+    const normals: Vector3[] = new Array(totalVertexCount);
+    const uvs: Vector2[] = new Array(totalVertexCount);
+
+    let indicesOffset = 0;
+    // create torso
+    for (let i = 0; i < torsoVertexCount; ++i) {
+      const x = i % radialCount;
+      const y = (i * radialCountReciprocal) | 0;
+      const u = x * radialSegmentsReciprocal;
+      const v = y * heightSegmentsReciprocal;
+      const theta = torsoThetaStart + u * torsoThetaRange;
+      const sinTheta = Math.sin(theta);
+      const cosTheta = Math.cos(theta);
+
+      let posX = radius * sinTheta;
+      let posY = y * unitHeight - halfHeight;
+      let posZ = radius * cosTheta;
+
+      // Position
+      positions[i] = new Vector3(posX, posY, posZ);
+      // Normal
+      normals[i] = new Vector3(sinTheta, 0, cosTheta);
+      // Texcoord
+      uvs[i] = new Vector2(u, 1 - v);
+    }
+
+    for (let i = 0; i < torsoRectangleCount; ++i) {
+      const x = i % radialSegments;
+      const y = (i * radialSegmentsReciprocal) | 0;
+
+      const a = y * radialCount + x;
+      const b = a + 1;
+      const c = a + radialCount;
+      const d = c + 1;
+
+      indices[indicesOffset++] = b;
+      indices[indicesOffset++] = c;
+      indices[indicesOffset++] = a;
+      indices[indicesOffset++] = b;
+      indices[indicesOffset++] = d;
+      indices[indicesOffset++] = c;
+    }
+
+    // top cap
+    for (let i = 0; i < capVertexCount; ++i) {
+      const x = i % radialCount;
+      const y = (i * radialCountReciprocal) | 0;
+      const u = x * radialSegmentsReciprocal;
+      const v = y * radialSegmentsReciprocal;
+      const alphaDelta = u * capTopAlphaRange;
+      const thetaDelta = v * capThetaRange;
+      const sinTheta = Math.sin(thetaDelta);
+
+      let posX = -radius * Math.cos(alphaDelta) * sinTheta;
+      let posY = radius * Math.cos(thetaDelta) + halfHeight;
+      let posZ = radius * Math.sin(alphaDelta) * sinTheta;
+
+      // Position
+      positions[i + torsoVertexCount] = new Vector3(posX, posY, posZ);
+      // Normal
+      normals[i + torsoVertexCount] = new Vector3(posX, posY, posZ);
+      // Texcoord
+      uvs[i + torsoVertexCount] = new Vector2(u, v);
+    }
+
+    for (let i = 0; i < capRectangleCount; ++i) {
+      const x = i % radialSegments;
+      const y = (i * radialSegmentsReciprocal) | 0;
+
+      const a = y * radialCount + x + torsoVertexCount;
+      const b = a + 1;
+      const c = a + radialCount;
+      const d = c + 1;
+
+      indices[indicesOffset++] = b;
+      indices[indicesOffset++] = a;
+      indices[indicesOffset++] = d;
+      indices[indicesOffset++] = a;
+      indices[indicesOffset++] = c;
+      indices[indicesOffset++] = d;
+    }
+
+    // bottom cap
+    for (let i = 0; i < capVertexCount; ++i) {
```

Comment:
bottom cap is similar to top cap, abstracting a function to set these buffer is better.